zerver/lib/actions.py

```diff
@@ -1305,14 +1305,31 @@ def do_add_reaction(user_profile, message, emoji_name):
     reaction.save()
     notify_reaction_update(user_profile, message, reaction, "add")
 
-def do_remove_reaction(user_profile, message, emoji_name):
+def do_remove_reaction_legacy(user_profile, message, emoji_name):
     # type: (UserProfile, Message, Text) -> None
     reaction = Reaction.objects.filter(user_profile=user_profile,
                                        message=message,
                                        emoji_name=emoji_name).get()
     reaction.delete()
     notify_reaction_update(user_profile, message, reaction, "remove")
 
+def do_add_reaction(user_profile, message, emoji_name, emoji_code, reaction_type):
+    # type: (UserProfile, Message, Text, Text, Text) -> None
+    reaction = Reaction(user_profile=user_profile, message=message,
+                        emoji_name=emoji_name, emoji_code=emoji_code,
+                        reaction_type=reaction_type)
+    reaction.save()
+    notify_reaction_update(user_profile, message, reaction, "add")
+
+def do_remove_reaction(user_profile, message, emoji_code, reaction_type):
+    # type: (UserProfile, Message, Text, Text) -> None
+    reaction = Reaction.objects.filter(user_profile=user_profile,
+                                       message=message,
+                                       emoji_code=emoji_code,
+                                       reaction_type=reaction_type).get()
```

Comment:
I guess the old version did this too, but I think this `.get()` is totally unnecessary, unless we want it to throw a 500 if there wasn't one previously (we probably want a JsonableError in that case instead).

Reply:
Agree. Updated! We already check and raise a JsonableError:
https://github.com/zulip/zulip/pull/7026/files#diff-fdedbbf3a7bb4156d79e6fac0735efa2R124

Reply:
Ahh! I had read it in hurry, we need `get()` as `notify_reaction_update()` uses the reaction instance.

Reply:
Tests are failing due to the removal of `send_message()` from `ZulipTestCase`, will fix tomorrow.